text/0000-tracked-properties.md

```diff
@@ -0,0 +1,1029 @@
+
+person.lastName = 'Jackson';
+```
+
+Now `fullName` knows which tracked properties were accessed when it was run,
+and setting `lastName` has invalidated `fullName`.
+
+> **NOTE:** This does _not_ invalidate a cache like in computed properties. Even
+> if `firstName` and `lastName` were untracked, the tracked getter would still
+> return the correct value on subsequent accesses, because `@tracked` does _not_
+> cache values. The validation and invalidation is pure meta data that is only
+> accessible by the Glimmer VM.
+>
+> Internally, Glimmer checks to see if a value has updated _before calling the
+> getter_. If it hasn't, then Glimmer does not rerender the related section of
+> the DOM. This is effectively an automatic `shouldComponentUpdate` from React.
+>
+> To prevent inconsistency, during development time, tracked properties will
+> keep a cache of their previous value to compare when they are activated and
+> ensure that it hasn't changed without invalidation. This will prevent improper
+> usage of tracked properties _outside_ of Glimmer's change tracking.
+
+### Manual Invalidation
+
+In user code, the idea that all mutable properties should be marked as tracked
+and that all other properties are effectively immutable works well in isolation.
+However, there are cases where users will want to work with code they do _not_
+control, such as external library code.
+
+Consider the following example. We have a `simple-timer` library that we've
+imported from NPM, and we're trying to wrap it with a `TimerComponent` that
+uses it to keep track of how much time has passed:
+
+```js
+// simple-timer/index.js
+export default class Timer {
+  seconds = 0;
+  minutes = 0;
+  hours = 0;
+
+  listeners = [];
+
+  constructor() {
+    setInterval(() => {
+      this.seconds++;
+      this.minutes = Math.floor(this.seconds / 60);
+      this.hours = Math.floor(this.minutes / 60);
+      this.notifyTick();
+    }, 1000);
+  }
+
+  notifyTick() {
+    for (let listener of this.listeners) {
+      listener(this.seconds);
+    }
+  }
+
+  onTick(listener) {
+    this.listeners.push(listener);
+  }
+}
+```
+
+```js
+import Timer from 'simple-timer';
+import Component, { tracked } from '@glimmer/tracking';
+
+export default class TimerComponent extends Component {
+  @tracked timer = new Timer();
+
+  @tracked
+  get currentSeconds() {
+    return this.timer.seconds;
+  }
+
+  @tracked
+  get currentMinutes() {
+    return this.timer.minutes;
+  }
+}
+```
+
+Even though we've marked the `timer` property as tracked, the `timer.seconds`
+property is untracked, and _it_ is the field that is updated. We can solve this
+problem by using the timer library's `onTick` event handler to re-set the field,
+invalidating it:
+
+```js
+export default class TimerComponent extends Component {
+  @tracked timer = new Timer();
+
+  constructor() {
+    this.timer.onTick(() => {
+      // invalidate the timer field.
+      this.timer = this.timer;
```

Comment:
Won't this invalidate the `@tracked` `currentMinutes()` ~59 times more than necessary? This seems rather excessive – are there any mechanisms to limit that impact?

Reply:
Glimmer’s architecture is designed to make this very, very cheap. I’d estimate the performance impact of this to be under 1ms per invalidation on almost every device—probably not something you could detect when profiling.

Reply:
I'm confused. In https://github.com/emberjs/rfcs/pull/410#discussion_r239590279 @pzuraq wants to limit re-renders, so it can't really be as inconsequential as you suggest?

People will likely make code like this, where it makes a significant difference. Eg. a 10.000 element array where a single element is changed every frame. Hence my question on how this can be mitigated?

Reply:
@kanongil what I meant in that comment was that we would be rerendering _the entire app_ in every render. This would be really expensive, but an individual tracked property is much much cheaper, which is why we want to avoid everything being volatile by default. Within Ember apps, this is likely not going to be an issue. If you updated a 10,000 item array that is rendered, Glimmer will intelligently only update the items that changed within that array. If you had a tracked property based on that array, it was going to rerun anyways, and that code is the responsibility of the user.

However, it is definitely true that users could experience issues if they do not have granularity when trying to wrap external libraries. For now, I think that ideally they would wrap the external library and use its hooks to get as granular as they can. In the future, I do think we should add more granular, low-level APIs which allow users to instrument external libraries more thoroughly. 

FWIW, this situation is not much different than where we are today with `get`/`set`, `notifyPropertyChange`, and computed properties. Wrapping external libraries is typically calling `notifyPropertyChange` on a computed from the libraries hooks, so the perf should be similar.